### PR TITLE
[1/4] Timer refactor: system timer

### PR DESCRIPTION
**Pull request by MabezDev**
## Thank you for your contribution!

We appreciate the time and effort you've put into this pull request.
To help us review it efficiently, please ensure you've gone through the following checklist:

### Submission Checklist 📝
- [x] I have updated existing examples or added new ones (if applicable).
- [x] I have used `cargo xtask fmt-packages` command to ensure that all changed code is formatted correctly.
- [x] My changes were added to the [`CHANGELOG.md`](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/CHANGELOG.md) in the **_proper_** section.
- [x] I have added necessary changes to user code to the [Migration Guide](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/MIGRATING-0.21.md).
- [x] My changes are in accordance to the [esp-rs API guidelines](https://github.com/esp-rs/esp-hal/blob/main/documentation/API-GUIDELINES.md)

#### Extra:
- [x] I have read the [CONTRIBUTING.md guide](https://github.com/esp-rs/esp-hal/blob/main/documentation/CONTRIBUTING.md) and followed its instructions.

#### Description

A fairly radical change, but I realized that it's basically impossible to treat the `Unit`'s as a resource _and_ allow users to actually make modifications (`FrozenUnit` prevented this before). `Unit` operations now go through the `SystemTimer` struct, and are all unsafe with safety comments explaining to the user how to use it correctly. Opening early for feedback before I update all the examples.

TL;DR: 

* Removes _all_ type params on Alarm
* Systimer no longer implements peripheral ref, the peripheral ref
  pattern is instead intended to be used on the higher level timer
  drivers (OneShot, Periodic)
* Removed `Unit` as a type, in favour of an enum
* Alarms are back in the main `SystemTimer` "driver"
* Made all `Unit` modification methods unsafe, it's not possible to
  modify the `Unit`'s safely whilst timers and or the `time::now` API is
  in use

### CODE REVIEW

**Comment by bjoernQ** [comment]
I welcome the simplification here especially since we are encouraging users to just use the higher-level timers and use the hw-timers to back them

I don't think users did anything but use `split` before in which case all the generics were just annoying best case.

Not sure if even the now unsafe functionality is useful, but it probably doesn't hurt to have it (while it would hurt to use it wrong - that's why it's unsafe)

**Comment by bjoernQ** [comment]
Unrelated but I think the `cycles` function in `examples/src/lib.rs` is unused and t.b.h. it should - the examples should be self contained (I remember the hate the example-simplification-macros in esp-wifi got back then)

File: examples/src/bin/hmac.rs [hunk 1]
```diff
@@ -94,18 +94,18 @@ fn main() -> ! {
         let mut remaining = nsrc;
         hw_hmac.init();
         block!(hw_hmac.configure(HmacPurpose::ToUser, KeyId::Key0)).expect("Key purpose mismatch");
-        let pre_hw_hmac = SystemTimer::now();
+        let pre_hw_hmac = SystemTimer::now(Default::default());
```
Comment (jessebraham) [inline examples/src/bin/hmac.rs:97]: Not a huge fan of this honestly, this is pretty perplexing without looking into the driver

Comment (MabezDev) [reply examples/src/bin/hmac.rs:97]: I wasn't too sure either, would you prefer 

- now(Unit::Unit0)
- now() as it was with a new method for accessing unit1's counter

**Comment by MabezDev** [comment]
Another question I'm not sure about, should Alarm's be ZSTs? I.e I erased the alarm number to simplify AnyTimer, but maybe that was a step too far for a "dumb timer"?

File: examples/src/bin/hmac.rs (hunk 1 above)
Comment (bjoernQ) [reply examples/src/bin/hmac.rs:97]: I personally would like just `now()`  better - maybe we could have another function which takes a Unit?

Comment (bugadani) [reply examples/src/bin/hmac.rs:97]: I'd rename this to `current_timestamp`, keep the unit. We already have `esp_hal::time::now()` and all other nows are rather confuing IMO.

Comment (MabezDev) [reply examples/src/bin/hmac.rs:97]: I've gone with `SystemTimer::unit_count(Unit)`, I plan on removing all the places in examples where we use `SystemTimer::now` for timing

File: esp-hal/src/timer/systimer.rs [hunk 2]
```diff
@@ -146,90 +93,81 @@ impl<'d> SystemTimer<'d> {
...
+            alarm0: Alarm::new(0),
+            alarm1: Alarm::new(1),
+            alarm2: Alarm::new(2),
         }
     }
 
-    /// Get the current count of Unit 0 in the System Timer.
-    pub fn now() -> u64 {
+    /// Get the current count of the given unit in the System Timer.
+    pub fn unit_count(unit: Unit) -> u64 {
```
Comment (bugadani) [inline esp-hal/src/timer/systimer.rs:111]: Counter, please, otherwise I'd expect this to return 2 :D

Comment (MabezDev) [reply esp-hal/src/timer/systimer.rs]: I went with `unit_value` instead, counter didn't feel correct either.

**Comment by MabezDev** [comment]
Skip changelog for embassy and esp-wifi changes, whilst technically breaking, they're both tied to esp-hal releases so it won't be breaking when they upgrade

**Review by bjoernQ — APPROVED** [review]
LGTM

**Review by playfulFence — APPROVED** [review]
Thanks!

File: esp-hal/src/timer/systimer.rs [hunk 3]
```diff
@@ -146,90 +93,81 @@ impl<'d> SystemTimer<'d> {
...
-            alarm2: Alarm::new_async(self.comparator2.into(), &unit),
-            #[cfg(not(esp32s2))]
-            unit1: self.unit1,
-        }
+    /// This can be used to load back the sleep time recorded by RTC timer
+    /// via software after Light-sleep
+    ///
+    /// # Safety
+    ///
+    /// - Modifying a unit's count whilst [`Alarm`]'s are using it may cause
```
Comment (jessebraham) [inline esp-hal/src/timer/systimer.rs:141]:
```suggestion
    /// - Modifying a unit's count whilst [`Alarm`]s are using it may cause
```

File: esp-hal/src/timer/systimer.rs [hunk 4]
```diff
@@ -332,54 +264,26 @@ pub trait Unit {
...
+        Alarm {
+            comp,
+            unit: Unit::Unit0,
         }
     }
-}
 
-/// A comparator that can generate alarms/interrupts based on values of a unit.
-pub trait Comparator {
     /// Returns the comparators number.
```
Comment (bugadani) [inline esp-hal/src/timer/systimer.rs:282]:
```suggestion
    /// Returns the comparator's number.
```

I know it's preexisting, but we can fix it 🙃

File: esp-hal/src/timer/systimer.rs [hunk 5]
```diff
@@ -431,12 +335,12 @@ pub trait Comparator {
...
     /// Sets the unit this comparator uses as a reference count.
     #[cfg(not(esp32s2))]
-    fn set_unit(&self, is_unit0: bool) {
+    pub fn set_unit(&self, unit: Unit) {
         let tconf = unsafe {
             let systimer = &*SYSTIMER::ptr();
             systimer.target_conf(self.channel() as usize)
         };
-        tconf.modify(|_, w| w.timer_unit_sel().bit(is_unit0));
+        tconf.modify(|_, w| w.timer_unit_sel().bit(matches!(unit, Unit::Unit0)));
```
Comment (bugadani) [inline esp-hal/src/timer/systimer.rs:343]: I believe this is backwards

**Review by bugadani — CHANGES_REQUESTED** [review]

File: esp-hal/src/timer/systimer.rs [hunk 6]
```diff
@@ -146,90 +93,81 @@ impl<'d> SystemTimer<'d> {
...
-        SysTimerAlarms {
-            alarm0: Alarm::new(self.comparator0.into(), &unit),
-            alarm1: Alarm::new(self.comparator1.into(), &unit),
-            alarm2: Alarm::new(self.comparator2.into(), &unit),
-            #[cfg(not(esp32s2))]
-            unit1: self.unit1,
-        }
+    /// # Safety
+    ///
+    /// - Disabling an `Unit` whilst [`Alarm`]'s are using it will affect the
```
[the remark this answers is not in the document]
Comment (bugadani) [reply esp-hal/src/timer/systimer.rs:126]: Also shouldn't it be "a Unit"? 🤔

File: esp-hal/src/timer/systimer.rs [hunk 7]
```diff
@@ -146,90 +93,81 @@ impl<'d> SystemTimer<'d> {
...
 /// A
 #[cfg_attr(esp32s2, doc = "64-bit")]
 #[cfg_attr(not(esp32s2), doc = "52-bit")]
 /// counter.
-pub trait Unit {
-    /// Returns the unit number.
-    fn channel(&self) -> u8;
+#[derive(Copy, Clone, Debug, Default, PartialEq, Eq)]
+pub enum Unit {
+    #[default]
```
Comment (bugadani) [inline esp-hal/src/timer/systimer.rs:156]: Considering that touching Unit0 can mess up the time keeping, do we really want to have a default?

File: esp-hal/src/timer/systimer.rs [hunk 8]
```diff
@@ -949,27 +589,30 @@ where
...
 }
 
+impl crate::private::Sealed for Alarm {}
+
 static CONF_LOCK: Lock = Lock::new();
 static INT_ENA_LOCK: Lock = Lock::new();
 
 // Async functionality of the system timer.
 mod asynch {
+    #![allow(unused)] // FIXME (mabez)
```
[the remark this answers is not in the document]
Comment (MabezDev) [reply esp-hal/src/timer/systimer.rs:615]: It's fixed in part 3 :)

File: esp-hal/src/timer/systimer.rs [hunk 9]
```diff
@@ -496,18 +400,6 @@ pub trait Comparator {
         }
     }
 
-    /// Get the actual target value of the comparator.
-    fn actual_target(&self) -> u64 {
```
[the remark this answers is not in the document]
Comment (MabezDev) [reply esp-hal/src/timer/systimer.rs:500]: We probably want to make this part of the `Timer` trait instead of an inherent method on alarm, I'm conscious in this PR to only expose systimer specific things publicly

**Review by bugadani — APPROVED** [review]
Thanks
